File: system-test/java/src/main/java/com/mycompany/myshop/testkit/driver/adapter/ui/client/pages/BasePage.java
package com.mycompany.myshop.testkit.driver.adapter.ui.client.pages;

import com.mycompany.myshop.testkit.driver.port.dtos.error.SystemError;
import com.mycompany.myshop.testkit.common.Result;
import com.mycompany.myshop.testkit.driver.adapter.shared.client.playwright.PageClient;

import java.util.List;

import static com.mycompany.myshop.testkit.dsl.core.usecase.commons.SystemResults.failure;
import static com.mycompany.myshop.testkit.dsl.core.usecase.commons.SystemResults.success;

public abstract class BasePage {
    private static final String NOTIFICATION_SELECTOR = "[role='alert']";
    private static final String NOTIFICATION_SUCCESS_SELECTOR = "[role='alert'].notification.success";
    private static final String NOTIFICATION_ERROR_SELECTOR = "[role='alert'].notification.error";
    private static final String NOTIFICATION_ERROR_MESSAGE_SELECTOR = "[role='alert'].notification.error .error-message";
    private static final String NOTIFICATION_ERROR_FIELD_SELECTOR = "[role='alert'].notification.error .field-error";
    private static final String NOTIFICATION_ID_ATTRIBUTE = "data-notification-id";

    private static final String UNRECOGNIZED_NOTIFICATION_ERROR_MESSAGE = "Notification type is not recognized";

    protected final PageClient pageClient;

    private String lastNotificationId = null;

    protected BasePage(PageClient pageClient) {
        this.pageClient = pageClient;
    }

    public Result<String, SystemError> getResult() {
        var notificationId = waitForNewNotification();
        lastNotificationId = notificationId;

        var isSuccess = isSuccessNotification(notificationId);

        if (isSuccess) {
            var successMessage = readSuccessNotification(notificationId);
            return success(successMessage);
        }

        var generalMessage = readGeneralErrorMessage(notificationId);
        var fieldErrorTexts = readFieldErrors(notificationId);

        if (fieldErrorTexts.isEmpty()) {
            return failure(generalMessage);
        }

        var fieldErrors = fieldErrorTexts.stream()
            .map(this::parseFieldError)
                .toList();

        var error = SystemError.builder()
                .message(generalMessage)
                .fields(fieldErrors)
                .build();

        return failure(error);
    }

    private String waitForNewNotification() {
        var selector = lastNotificationId == null
                ? NOTIFICATION_SELECTOR + "[" + NOTIFICATION_ID_ATTRIBUTE + "]"
                : NOTIFICATION_SELECTOR + "[" + NOTIFICATION_ID_ATTRIBUTE + "]:not([" + NOTIFICATION_ID_ATTRIBUTE + "='" + lastNotificationId + "'])";

        pageClient.waitForVisible(selector);

        var notificationId = pageClient.readAttribute(selector, NOTIFICATION_ID_ATTRIBUTE);

        if(notificationId == null) {
            throw new IllegalStateException("Notification does not have an ID");
        }

        return notificationId;
    }

    private boolean isSuccessNotification(String notificationId) {
        var successSelector = withNotificationId(NOTIFICATION_SUCCESS_SELECTOR, notificationId);
        var isSuccess = pageClient.isVisible(successSelector);

        if(isSuccess) {
            return true;
        }

        var errorSelector = withNotificationId(NOTIFICATION_ERROR_SELECTOR, notificationId);
        var isError = pageClient.isVisible(errorSelector);

        if(isError) {
            return false;
        }

        throw new IllegalStateException(UNRECOGNIZED_NOTIFICATION_ERROR_MESSAGE);
    }

    private String readSuccessNotification(String notificationId) {
        var selector = withNotificationId(NOTIFICATION_SUCCESS_SELECTOR, notificationId);
        return pageClient.readTextContent(selector);
    }

    private String readGeneralErrorMessage(String notificationId) {
        var selector = withNotificationId(NOTIFICATION_ERROR_MESSAGE_SELECTOR, notificationId);
        return pageClient.readTextContent(selector);
    }

    private List<String> readFieldErrors(String notificationId) {
        var selector = withNotificationId(NOTIFICATION_ERROR_FIELD_SELECTOR, notificationId);
        if (!pageClient.isVisible(selector)) {
            return List.of();
        }
        return pageClient.readAllTextContents(selector);
    }

    private SystemError.FieldError parseFieldError(String text) {
        var parts = text.split(":", 2);

        if (parts.length != 2) {
            throw new IllegalArgumentException("Invalid field error format: " + text);
        }

        return new SystemError.FieldError(
                parts[0].trim(),
                parts[1].trim()
        );
    }

    private static String withNotificationId(String selector, String notificationId) {
        var idAttribute = "[" + NOTIFICATION_ID_ATTRIBUTE + "='" + notificationId + "']";
        return selector.replace(NOTIFICATION_SELECTOR, NOTIFICATION_SELECTOR + idAttribute);
    }
}

File: system-test/dotnet/Driver.Adapter/Ui/Client/Pages/BasePage.cs
using Driver.Adapter.Shared.Client.Playwright;

using Common;

using Driver.Port;

using Driver.Port.Dtos.Error;



namespace Driver.Adapter.Ui.Client.Pages;



public abstract class BasePage

{

    private const string NotificationSelector = "[role='alert'][data-notification-id]";

    private const string NotificationSuccessSelector = "[role='alert'].notification.success";

    private const string NotificationErrorSelector = "[role='alert'].notification.error";

    private const string NotificationErrorMessageSelector = "[role='alert'].notification.error .error-message";

    private const string NotificationErrorFieldSelector = "[role='alert'].notification.error .field-error";

    private const string NotificationIdAttribute = "data-notification-id";

    private const string UnrecognizedNotificationErrorMessage = "Notification type is not recognized";



    protected readonly PageClient PageClient;

    

    private string? _lastNotificationId;



    protected BasePage(PageClient pageClient)

    {

        PageClient = pageClient;

    }



    public async Task<Result<string, SystemError>> GetResultAsync()

    {

        var notificationId = await WaitForNewNotificationAsync();

        _lastNotificationId = notificationId;



        var isSuccess = await IsSuccessNotificationAsync(notificationId);



        if (isSuccess)

        {

            var successMessage = await ReadSuccessNotificationAsync(notificationId);

            return SystemResults.Success(successMessage);

        }



        var generalMessage = await ReadGeneralErrorMessageAsync(notificationId);

        var fieldErrorTexts = await ReadFieldErrorsAsync(notificationId);



        if (fieldErrorTexts.Count == 0)

        {

            return SystemResults.Failure<string>(generalMessage);

        }



        var fieldErrors = fieldErrorTexts.Select(ParseFieldError).ToList();



        var error = SystemError.Of(generalMessage, fieldErrors);



        return SystemResults.Failure<string>(error);

    }



    private async Task<string> WaitForNewNotificationAsync()

    {

        var selector = _lastNotificationId == null

            ? NotificationSelector

            : $"{NotificationSelector}:not([{NotificationIdAttribute}='{_lastNotificationId}'])";



        await PageClient.WaitForVisibleAsync(selector);



        var notificationId = await PageClient.ReadAttributeAsync(selector, NotificationIdAttribute);

        

        if(notificationId == null) 

        {

            throw new InvalidOperationException($"Notification element does not have {NotificationIdAttribute} attribute");

        }



        return notificationId;

    }



    private async Task<bool> IsSuccessNotificationAsync(string notificationId)

    {

        var successSelector = WithNotificationId(NotificationSuccessSelector, notificationId);

        var isSuccess = await PageClient.IsVisibleAsync(successSelector);



        if (isSuccess)

        {

            return true;

        }



        var errorSelector = WithNotificationId(NotificationErrorSelector, notificationId);

        var isError = await PageClient.IsVisibleAsync(errorSelector);



        if (isError)

        {

            return false;

        }



        throw new InvalidOperationException(UnrecognizedNotificationErrorMessage);

    }



    private async Task<string> ReadSuccessNotificationAsync(string notificationId)

    {

        var selector = WithNotificationId(NotificationSuccessSelector, notificationId);

        return await PageClient.ReadTextContentAsync(selector);

    }



    private async Task<string> ReadGeneralErrorMessageAsync(string notificationId)

    {

        var selector = WithNotificationId(NotificationErrorMessageSelector, notificationId);

        return await PageClient.ReadTextContentAsync(selector);

    }



    private async Task<List<string>> ReadFieldErrorsAsync(string notificationId)

    {

        var selector = WithNotificationId(NotificationErrorFieldSelector, notificationId);



        if (!await PageClient.IsVisibleAsync(selector))

        {

            return new List<string>();

        }

        

        return await PageClient.ReadAllTextContentsAsync(selector);

    }



    private static SystemError.FieldError ParseFieldError(string text)

    {

        var parts = text.Split(':', 2);



        if (parts.Length != 2)

        {

            throw new InvalidOperationException($"Invalid field error format: {text}");

        }



        return new SystemError.FieldError(parts[0].Trim(), parts[1].Trim());

    }



    private static string WithNotificationId(string selector, string notificationId)

    {

        var idAttribute = $"[{NotificationIdAttribute}='{notificationId}']";

        return selector.Replace(NotificationSelector, NotificationSelector + idAttribute);

    }

}







